File: packages/ui/src/parking-picker.test.ts
import { describe, expect, test } from "bun:test";
import {
  assignableParkingSlots,
  parkingKindOf,
  parkingNumberHint,
  parkingSlotLabel,
  preferredParkingKind,
} from "./parking-picker";

const openFree = {
  id: "open-1",
  kind: "open" as const,
  wing: null,
  slotNumber: "12",
  flatId: null,
  flatNumber: null,
};
const puzzleTaken = {
  id: "puz-1",
  kind: "puzzle" as const,
  wing: "A",
  slotNumber: "101",
  flatId: "other-flat",
  flatNumber: "202",
};
const puzzleMine = {
  id: "puz-2",
  kind: "puzzle" as const,
  wing: "A",
  slotNumber: "102",
  flatId: "flat-1",
  flatNumber: "101",
};

describe("parking-picker", () => {
  test("treats missing kind as open so older lots still list", () => {
    expect(parkingKindOf({ kind: "puzzle" })).toBe("puzzle");
    expect(parkingKindOf({ kind: "open" })).toBe("open");
    expect(parkingKindOf({ kind: "" as "open" })).toBe("open");
  });

  test("labels puzzle as wing · number", () => {
    expect(parkingSlotLabel(puzzleTaken)).toBe("A · 101");
    expect(parkingSlotLabel(openFree)).toBe("12");
  });

  test("hides lots already assigned to another flat", () => {
    const rows = [openFree, puzzleTaken, puzzleMine];
    expect(
      assignableParkingSlots(rows, "puzzle", "flat-1", "").map((p) => p.id),
    ).toEqual(["puz-2"]);
    expect(
      assignableParkingSlots(rows, "open", "flat-1", "").map((p) => p.id),
    ).toEqual(["open-1"]);
  });

  test("prefers the kind that actually has a free lot", () => {
    expect(preferredParkingKind([openFree], undefined, "flat-1", "")).toBe(
      "open",
    );
    expect(
      preferredParkingKind([openFree, puzzleMine], puzzleMine, "flat-1", "puz-2"),
    ).toBe("puzzle");
  });

  test("explains why Parking number is only None", () => {
    expect(parkingNumberHint([], "puzzle", "flat-1", "")).toContain(
      "Add parking lots in Manage",
    );
    expect(parkingNumberHint([openFree], "puzzle", "flat-1", "")).toContain(
      "Switch Parking type to Open",
    );
    expect(parkingNumberHint([puzzleTaken], "puzzle", "flat-1", "")).toContain(
      "assigned to other flats",
    );
    expect(parkingNumberHint([puzzleTaken, openFree], "puzzle", "flat-1", "")).toContain(
      "Switch to Open",
    );
    expect(preferredParkingKind([], undefined, "flat-1", "")).toBe("puzzle");
    expect(parkingNumberHint([puzzleTaken], "open", "flat-1", "")).toContain(
      "Switch Parking type to see Puzzle",
    );
  });
});

File: apps/mobile/test/unit/parking_picker_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:societyhub_mobile/api/models.dart';
import 'package:societyhub_mobile/shared/parking_picker.dart';

ParkingSlotDto _slot({
  required String id,
  required String slotNumber,
  required String kind,
  String? flatId,
  String? wing,
  String? flatNumber,
}) {
  return ParkingSlotDto(
    id: id,
    slotNumber: slotNumber,
    kind: kind,
    flatId: flatId,
    wing: wing,
    flatNumber: flatNumber,
  );
}

void main() {
  final openFree = _slot(id: 'open-1', slotNumber: '12', kind: 'open');
  final puzzleTaken = _slot(
    id: 'puz-1',
    slotNumber: '101',
    kind: 'puzzle',
    wing: 'A',
    flatId: 'other-flat',
    flatNumber: '202',
  );
  final puzzleMine = _slot(
    id: 'puz-2',
    slotNumber: '102',
    kind: 'puzzle',
    wing: 'A',
    flatId: 'flat-1',
    flatNumber: '101',
  );

  test('hides lots already assigned to another flat', () {
    final rows = [openFree, puzzleTaken, puzzleMine];
    expect(
      assignableParkingSlots(rows, 'puzzle', 'flat-1', null).map((p) => p.id),
      ['puz-2'],
    );
    expect(
      assignableParkingSlots(rows, 'open', 'flat-1', null).map((p) => p.id),
      ['open-1'],
    );
  });

  test('prefers the kind that actually has a free lot', () {
    expect(preferredParkingKind([openFree], null, 'flat-1', null), 'open');
    expect(
      preferredParkingKind([openFree, puzzleMine], puzzleMine, 'flat-1', 'puz-2'),
      'puzzle',
    );
  });

  test('explains why Parking number is only None', () {
    expect(
      parkingNumberHint([], 'puzzle', 'flat-1', null),
      contains('Add parking lots in Manage'),
    );
    expect(
      parkingNumberHint([openFree], 'puzzle', 'flat-1', null),
      contains('Switch Parking type to Open'),
    );
    expect(
      parkingNumberHint([puzzleTaken], 'puzzle', 'flat-1', null),
      contains('assigned to other flats'),
    );
  });
}
